File: android/app/src/main/java/com/singword/app/data/local/song/SongMatchSnapshot.kt
package com.singword.app.data.local.song

import com.singword.app.domain.MatchedWord

data class SongWordSnapshot(
    val word: String,
    val pos: String,
    val def: String,
    val source: String
)

data class SongMatchSnapshot(
    val id: String,
    val trackName: String,
    val artistName: String,
    val provider: String,
    val totalTokens: Int,
    val matchedWords: List<SongWordSnapshot>,
    val timestamp: Long = System.currentTimeMillis()
) {
    companion object {
        fun createId(trackName: String, artistName: String): String {
            return "${trackName.trim().lowercase()}|${artistName.trim().lowercase()}"
        }

        fun fromMatchedWords(
            trackName: String,
            artistName: String,
            provider: String,
            totalTokens: Int,
            matchedWords: List<MatchedWord>,
            timestamp: Long = System.currentTimeMillis()
        ): SongMatchSnapshot {
            return SongMatchSnapshot(
                id = createId(trackName, artistName),
                trackName = trackName,
                artistName = artistName,
                provider = provider,
                totalTokens = totalTokens,
                matchedWords = matchedWords.map {
                    SongWordSnapshot(
                        word = it.word,
                        pos = it.pos,
                        def = it.def,
                        source = it.source
                    )
                },
                timestamp = timestamp
            )
        }
    }

    fun toMatchedWords(): List<MatchedWord> = matchedWords.map {
        MatchedWord(
            word = it.word,
            pos = it.pos,
            def = it.def,
            source = it.source
        )
    }
}

File: ios/singword/Data/Songs/SongMatchSnapshot.swift
import Foundation

nonisolated struct SongWordSnapshot: Codable, Hashable, Identifiable, Sendable {
    var id: String { word }

    let word: String
    let pos: String
    let definition: String
    let source: String
}

nonisolated struct SongMatchSnapshot: Codable, Hashable, Identifiable, Sendable {
    let trackName: String
    let artistName: String
    let provider: String
    let totalTokens: Int
    let matchedWords: [SongWordSnapshot]
    let timestamp: TimeInterval

    var id: String {
        "\(trackName.lowercased())|\(artistName.lowercased())"
    }

    init(
        trackName: String,
        artistName: String,
        provider: String,
        totalTokens: Int,
        matchedWords: [SongWordSnapshot],
        timestamp: TimeInterval = Date().timeIntervalSince1970
    ) {
        self.trackName = trackName
        self.artistName = artistName
        self.provider = provider
        self.totalTokens = totalTokens
        self.matchedWords = matchedWords
        self.timestamp = timestamp
    }

    init(
        trackName: String,
        artistName: String,
        provider: String,
        totalTokens: Int,
        matchedWords: [MatchedWord],
        timestamp: TimeInterval = Date().timeIntervalSince1970
    ) {
        self.init(
            trackName: trackName,
            artistName: artistName,
            provider: provider,
            totalTokens: totalTokens,
            matchedWords: matchedWords.map {
                SongWordSnapshot(
                    word: $0.word,
                    pos: $0.pos,
                    definition: $0.definition,
                    source: $0.source
                )
            },
            timestamp: timestamp
        )
    }

    var matchedWordItems: [MatchedWord] {
        matchedWords.map {
            MatchedWord(
                word: $0.word,
                pos: $0.pos,
                definition: $0.definition,
                source: $0.source
            )
        }
    }
}
